sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/LLVMElemPtrSymbol.java

```diff
@@ -0,0 +1,91 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of
+ * conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of
+ * conditions and the following disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.oracle.truffle.llvm.runtime;
+
+import com.oracle.truffle.llvm.runtime.IDGenerater.BitcodeID;
+import com.oracle.truffle.llvm.runtime.global.LLVMGlobal;
+import com.oracle.truffle.llvm.runtime.nodes.memory.LLVMGetElementPtrNode;
+import com.oracle.truffle.llvm.runtime.types.Type;
+
+public final class LLVMElemPtrSymbol extends LLVMSymbol {
+    private final Type type;
+    private final LLVMGetElementPtrNode elementPtrNode;
+
+    public LLVMElemPtrSymbol(String name, BitcodeID bitcodeID, int symbolIndex, boolean exported, Type type, LLVMSymbol base, LLVMGetElementPtrNode elementPtrNode) {
+        super(name, bitcodeID, symbolIndex, exported, base.isExternalWeak());
+        this.type = type;
+        this.elementPtrNode = elementPtrNode;
+    }
+
+    @Override
+    public boolean isGlobalVariable() {
+        return false;
+    }
+
+    @Override
+    public boolean isFunction() {
+        return false;
+    }
+
+    @Override
+    public boolean isAlias() {
+        return false;
+    }
+
+    @Override
+    public LLVMFunction asFunction() {
+        throw new IllegalStateException("GetElementPointerConstant " + getName() + " has to be resolved and might not be a function.");
+    }
+
+    @Override
+    public LLVMGlobal asGlobalVariable() {
+        throw new IllegalStateException("GetElementPointerConstant " + getName() + " has to be resolved and might not be a global variable.");
+    }
+
+    public final Type getType() {
+        return type;
+    }
+
+    public final LLVMGetElementPtrNode getElementPtrNode() {
+        return elementPtrNode;
```

Comment:
Storing the node here and just returning it is a problem. This function is used in an `@Cached` specification. That means whenever this is used anywhere in the AST, this node is added as a child node. But nodes can only be added as a child node at a single point in the AST.

What if this symbol is used in more than one place?

I think it would be better to have a `createElementPtrNode` function in this class, that re-creates the GEP node on the fly for every individual user.